cunumeric/array.py

```diff
@@ -2040,6 +2049,58 @@ def diagonal(
                 raise ValueError("Either axis1/axis2 or axes must be supplied")
         return self._diag_helper(offset=offset, axes=axes, extract=extract)
 
+    def trace(self, offset=0, axis1=None, axis2=None, dtype=None, out=None):
+        """a.trace(offset=0, axis1=None, axis2=None, dtype = None, out = None)
+
+        Return the sum along diagonals of the array.
+
+        Refer to :func:`cunumeric.trace` for full documentation.
+
+        See Also
+        --------
+        cunumeric.trace : equivalent function
+
+        Availability
+        --------
+        Multiple GPUs, Multiple CPUs
+
+        """
+        if self.ndim < 2:
+            raise ValueError(
+                "trace operation can't be called on a array with DIM<2"
+            )
+
+        axes = []
+        if (axis1 is None) and (axis2 is None):
+            # default values for axis
+            axes = (0, 1)
+        elif (axis1 is None) or (axis2 is None):
+            return ValueError("both axis should be passed")
+        else:
+            axes = (axis1, axis2)
+
+        res = self._diag_helper(offset=offset, axes=axes, trace=True)
+
+        # for 2D arrays we must return scalar
+        if self.ndim == 2:
+            res = res[0]
+
+        if out is not None:
+            out = convert_to_cunumeric_ndarray(out)
+            # out should be right shape
+            if out.shape != res.shape:
+                return ValueError("out has a wrong shape")
+            # type of out should be preserved
+            if out.dtype != res.dtype:
+                out._thunk.convert(res._thunk)
+            else:
+                out._thunk.copy(res._thunk)
```

Comment:
Could you instead pass the `out` array to `_diag_helper`, and reuse it if possible? That way we can save this copy if the `out` array is already the correct shape and type.

Reply:
Sounds good. Will do that